src/table.rs

```diff
@@ -280,26 +350,230 @@ where
+            }
+            1.0 / suffix_count as f64
+        };
 
-    /// Checks if the suffix table is lexicographically sorted. This is always true for valid suffix tables.
-    pub fn is_sorted(&self) -> bool {
-        self.table
-            .windows(2)
-            .all(|pair| self.text[pair[0] as usize..] <= self.text[pair[1] as usize..])
+        let (gt_zero_count, eq_one_count) = self.get_occurrence_counts(&counts);
+        let used_suffix_count = gt_zero_count as f64;
+        let used_once_suffix_count = eq_one_count as f64;
+
+        // Interpolation budget to be distributed according to lower order n-gram distribution
+        let delta = self.get_cached_delta(query.len() + 1);
+        let lambda = if delta < 1.0 {
+            delta.mul(used_suffix_count).mul(suffix_count_recip)
+        } else {
+            used_once_suffix_count
+                + delta
+                    .mul(used_suffix_count - used_once_suffix_count)
+                    .mul(suffix_count_recip)
+        };
+        
+        let mut probs = Vec::with_capacity(counts.len());
+        counts
+            .iter()
+            .zip(p_continuations.iter())
+            .for_each(|(&count, &p_continuation)| {
+                let prob =
+                    (count as f64 - delta).max(0.0).mul(suffix_count_recip) + lambda.mul(p_continuation);
+                probs.push(prob);
+            });
+        probs
+    }
+
+    /// Returns tuple of the number of elements that are greater than 0 and the number of elements that equal 1.
+    fn get_occurrence_counts(&self, slice: &[usize]) -> (usize, usize) {
+        slice
+            .iter()
+            .fold((0, 0), |(gt_zero_count, eq_one_count), &c| {
+                let gt_zero_count = gt_zero_count + (c > 0) as usize;
+                let eq_one_count = eq_one_count + (c == 1) as usize;
+                (gt_zero_count, eq_one_count)
+            })
+    }
+
+    /// Autoregressively sample k characters from a Kneser-Ney smoothed n-gram model.
+    fn kn_sample(&self, query: &[u16], n: usize, k: usize, vocab: Option<u16>) -> Result<Vec<u16>> {
+        let mut rng = thread_rng();
+        let mut sequence = Vec::from(query);
+
+        for _ in 0..k {
+            let start = sequence.len().saturating_sub(n - 1);
+            let prev = &sequence[start..];
+            let probs = self.smoothed_probs(prev, vocab);
+            let dist = WeightedIndex::new(&probs)?;
+            let sampled_index = dist.sample(&mut rng);
+
+            sequence.push(sampled_index as u16);
+        }
+
+        Ok(sequence)
+    }
+
+    // For a given n, produce a map from an occurrence count in (1, 2) to the number of unique n-grams with that occurrence count.
+    fn count_ngrams(&self, n: usize) -> HashMap<usize, usize> {
+        let mut count_map: HashMap<usize, usize> = [(1, 0), (2, 0)]
+            .iter()
+            .cloned()
+            .collect();
+
+        let query = vec![0; 0];
```

Comment:
Wait isn't this just an empty Vec in which case you could use `Vec::new()` or just `vec![]`?